src/lib.rs

```diff
@@ -923,118 +901,52 @@ fn format_option(opt: &OptGroup) -> String {
     line
 }
 
-#[derive(Clone, Copy)]
-enum SplitWithinState {
-    A,  // leading whitespace, initial state
-    B,  // words
-    C,  // internal and trailing whitespace
-}
-
-#[derive(Clone, Copy)]
-enum Whitespace {
-    Ws, // current char is whitespace
-    Cr  // current char is not whitespace
-}
-
-#[derive(Clone, Copy)]
-enum LengthLimit {
-    UnderLim, // current char makes current substring still fit in limit
-    OverLim   // current char makes current substring no longer fit in limit
-}
-
-
 /// Splits a string into substrings with possibly internal whitespace,
 /// each of them at most `lim` bytes long, if possible. The substrings
 /// have leading and trailing whitespace removed, and are only cut at
 /// whitespace boundaries.
 ///
-/// Note: Function was moved here from `std::str` because this module is the only place that
-/// uses it, and because it was too specific for a general string function.
-fn each_split_within<'a, F>(ss: &'a str, lim: usize, mut it: F)
-                            -> bool where F: FnMut(&'a str) -> bool {
-    // Just for fun, let's write this as a state machine:
-
-    let mut slice_start = 0;
-    let mut last_start = 0;
-    let mut last_end = 0;
-    let mut state = A;
-    let mut fake_i = ss.len();
-    let mut lim = lim;
-
-    let mut cont = true;
-
-    // if the limit is larger than the string, lower it to save cycles
-    if lim >= fake_i {
-        lim = fake_i;
-    }
-
-    let mut machine = |cont: &mut bool, state: &mut SplitWithinState, (i, c): (usize, char)| {
-        let whitespace = if c.is_whitespace() { Ws }       else { Cr };
-        let limit      = if (i - slice_start + 1) <= lim  { UnderLim } else { OverLim };
-
-        *state = match (*state, whitespace, limit) {
-            (A, Ws, _)        => { A }
-            (A, Cr, _)        => { slice_start = i; last_start = i; B }
-
-            (B, Cr, UnderLim) => { B }
-            (B, Cr, OverLim)  if (i - last_start + 1) > lim => {
-                // A single word has gone over the limit.  In this
-                // case we just accept that the word will be too long.
-                B
+/// Note: Function was moved here from `std::str` because this module
+/// is the only place that uses it, and because it was too specific for
+/// a general string function.
+fn each_split_within(desc: &String, lim: usize) -> Vec<String> {
+    let mut rows = Vec::new();
+    for line in desc.trim().lines() {
+        let mut words = Vec::new();
+        let mut word = String::new();
```

Comment:
We've got a fair amount of allocation going on in this method that would be good to avoid if we can. It seems like we're processing the line multiple times to clear out excess whitespace. We could do this without the temporary strings by maintaining an index into the line that we're processing:

```rust
// Add an additional whitespace to flush the last word
let line_chars = line.chars().chain(Some(' '));

let words = line_chars.fold((Vec::new(), 0, 0), |(mut words, word_start_idx, last_idx), c| {
    // Get the current byte offset
    let idx = last_idx + c.len_utf8();

    // If the char is whitespace, advance the word start and maybe push a word
    if c.is_whitespace() {
        if word_start_idx != last_idx {
            words.push(&line[word_start_idx..last_idx]);
        }

        (words, idx, idx)
    }
    // If the char is not whitespace, continue, retaining the current
    else {
        (words, word_start_idx, idx)
    }
}).0;
```

The example uses the `Iterator::fold` method to let us thread state through our chars, so we can find the point at which a word start, then keep that index until we hit the end. Here's a [runnable version you can check out](http://play.integer32.com/?gist=a089f5ae469d9924aa8982d277b58d70&version=stable&mode=debug).

Reply:
Looks neat ! 👍